components/script/dom/bluetooth.rs

```diff
@@ -40,59 +60,163 @@ impl Bluetooth {
     }
 }
 
+impl Clone for BluetoothScanFilter {
+    fn clone(&self) -> BluetoothScanFilter {
+        BluetoothScanFilter {
+            name: self.name.clone(),
+            namePrefix: self.namePrefix.clone(),
+            services: self.services.clone(),
+        }
+    }
+}
+
+impl Clone for RequestDeviceOptions {
+    fn clone(&self) -> RequestDeviceOptions {
+        RequestDeviceOptions {
+            filters: self.filters.clone(),
+            optionalServices: self.optionalServices.clone(),
+        }
+    }
+}
+
+fn canonicalize_filter(filter: &BluetoothScanFilter, global: GlobalRef)
+    -> Fallible<BluetoothScanfilter> {
+    if !(filter.services.is_some() ||
+         filter.name.is_some() ||
+         filter.namePrefix.is_some()) {
+        return Err(Type(FILTER_ERROR.to_owned()));
+    }
+
+    let mut services_vec: Vec<String> = vec!();
+    if let Some(services) = filter.services.clone() {
+        if services.is_empty() {
+            return Err(Type(SERVICE_ERROR.to_owned()));
+        }
+        for service in services {
+            match BluetoothUUID::GetService(global, service) {
+                Ok(valid) => services_vec.push(valid.to_string()),
+                Err(err) => return Err(err),
+            }
+        }
+    }
+
+    let mut name = String::new();
+    if let Some(filter_name) = filter.name.clone() {
+        //NOTE: DOMString::len() gives back the size in bytes
+        if filter_name.len() > MAX_DEVICE_NAME_LENGTH {
+            return Err(Type(NAME_TOO_LONG_ERROR.to_owned()));
+        }
+        if filter_name.len() > MAX_FILTER_NAME_LENGTH {
+            return Err(Type(FILTER_NAME_TOO_LONG_ERROR.to_owned()));
+        }
+        name = filter_name.to_string();
+    }
+
+    let mut name_prefix = String::new();
+    if let Some(filter_name_prefix) = filter.namePrefix.clone() {
```

Comment:
yes we need, because filter.namePrefix is a DOMString and DOMString does not have the Copy trait implemented, so we need explicitly write the .clone()